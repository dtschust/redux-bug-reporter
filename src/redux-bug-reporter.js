/* global fetch */
import React from 'react'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import isEqual from 'lodash.isequal'
import isFunction from 'lodash.isfunction'
import { fromJS } from 'immutable'
import { middlewareData, overloadStore, initializePlayback, finishPlayback, playbackFlag } from './store-enhancer'
import isClientRender from './is-client-render'
import { listenToErrors, errorData } from './utils'
require('es6-promise').polyfill()
require('isomorphic-fetch')

// On the server, UnconnectedBugReporter is a placeholder component
let UnconnectedBugReporter = () => {
  return (<span/>)
}

if (isClientRender()) {
  UnconnectedBugReporter = React.createClass({
    displayName: 'Bug Reporter',

    propTypes: {
      // passed in from parent
      submit: React.PropTypes.oneOfType([
        React.PropTypes.func,
        React.PropTypes.string
      ]).isRequired,
      projectName: React.PropTypes.string.isRequired,
      redactStoreState: React.PropTypes.func,
      name: React.PropTypes.string,
      meta: React.PropTypes.any,
      stringifyPayload: React.PropTypes.bool,
      // Passed in by redux-bug-reporter
      dispatch: React.PropTypes.func.isRequired,
      storeState: React.PropTypes.any.isRequired,
      overloadStore: React.PropTypes.func.isRequired,
      initializePlayback: React.PropTypes.func.isRequired,
      finishPlayback: React.PropTypes.func.isRequired
    },

    getInitialState: function () {
      return {
        expanded: false,
        mounted: false,
        loading: false,
        bugFiled: false,
        reporter: this.props.name || '',
        description: '',
        screenshotURL: '',
        notes: '',
        error: '',
        bugURL: ''
      }
    },

    shouldComponentUpdate: function (nextProps, nextState) {
      // Do not bother rerendering every props change.
      // Rerender only needs to occur on state change
      if (this.state !== nextState) {
        return true
      }
      return false
    },

    componentDidMount: function () {
      this.setState({ mounted: true })
      listenToErrors()
      // Global function to play back a bug
      window.bugReporterPlayback = this.bugReporterPlayback
    },

    toggleExpanded: function () {
      this.setState({expanded: !this.state.expanded})
    },

    bugReporterPlayback: function (actions, initialState, finalState, options = { delay: 100, immutable: false }) {
      let { dispatch, overloadStore } = this.props
      let { delay, immutable } = options

      if (delay === -1) {
        // Do not playback, just jump to the final state
        overloadStore((immutable ? fromJS(finalState) : finalState))
        return
      }

      this.props.initializePlayback((immutable ? fromJS(initialState) : initialState))
      overloadStore()

      const performNextAction = () => {
        let action = actions[0]

        // Let store know this is a playback action
        action[playbackFlag] = true

        dispatch(action)
        actions.splice(0, 1)
        if (actions.length > 0) {
          setTimeout(performNextAction, delay)
        } else {
          this.props.finishPlayback()
          let storeState = this.props.storeState
          let keys = Object.keys(storeState)
          keys.forEach(function (key) {
            if (!isEqual(storeState[key], finalState[key])) {
              console.log('The following reducer does not strictly equal the bug report final state: ' + key + '. I\'ll print them both out so you can see the differences.')
              console.log(key + ' current state:', storeState[key], '\n' + key + ' bug report state:', finalState[key])
            }
          })
          console.log('Playback complete!')
        }
      }

      performNextAction()
    },

    submit: function (e) {
      e.preventDefault()
      const {submit, projectName, storeState, redactStoreState, meta, stringifyPayload} = this.props
      let {reporter, description, screenshotURL, notes} = this.state
      this.setState({loading: true})

      let state = storeState
      let initialState = middlewareData.getBugReporterInitialState()
      let promise
      if (redactStoreState) {
        initialState = redactStoreState(initialState)
        state = redactStoreState(state)
      }
      const newBug = {
        projectName,
        state,
        initialState,
        actions: middlewareData.getActions(),
        consoleErrors: errorData.getErrors(),
        reporter,
        description,
        screenshotURL,
        notes,
        meta,
        useragent: window.navigator.userAgent,
        windowDimensions: [window.innerWidth, window.innerHeight],
        windowLocation: window.location.href
      }

      // Stringify payload if required, useful for filing via sheetsu
      if (stringifyPayload) {
        newBug.actions = JSON.stringify(newBug.actions)
        newBug.state = JSON.stringify(newBug.state)
        newBug.initialState = JSON.stringify(newBug.initialState)
      }

      // if submit is a function, call it instead of fetching
      // and attach to the promise returned
      if (isFunction(submit)) {
        promise = submit(newBug)
      } else {
        promise = fetch(submit, {
          method: 'post',
          headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
          },
          body: JSON.stringify(newBug)
        }).then(function (response) {
          return response.json()
        })
      }

      promise.then((json = {}) => {
        let {bugURL} = json
        this.setState({loading: false, bugFiled: true, bugURL, expanded: true})
      }).catch((error) => {
        console.error('Error filing bug', error)
        this.setState({loading: false, bugFiled: true, error, expanded: true})
      })
    },

    dismiss: function (e) {
      e.preventDefault()
      this.setState({bugFiled: false, expanded: false, bugURL: ''})
    },

    handleChange: function (field) {
      return (e) => {
        this.setState({ [field]: e.target.value })
      }
    },

    render: function () {
      let {reporter, description, screenshotURL, notes, mounted, loading, bugFiled, error, expanded, bugURL} = this.state
      if (!mounted) {
        return false
      }
      if (loading) {
        return loadingLayout
      }

      if (bugFiled) {
        return (
          <div className='Redux-Bug-Reporter'>
            <div className={`Redux-Bug-Reporter__form Redux-Bug-Reporter__form--${error ? 'fail' : 'success'}`}>
              {error ? (
                <div>
                  <div>Oops, something went wrong!</div>
                  <div>Please try again later</div>
                </div>
              ) : (
                <div>
                  <div>Your bug has been filed successfully!</div>
                  {bugURL && <div><a target='_blank' href={bugURL}>Here is a link to it!</a></div>}
                </div>
              )}
            </div>
            <div className='Redux-Bug-Reporter__show-hide-container'>
              <button className={`Redux-Bug-Reporter__show-hide-button Redux-Bug-Reporter__show-hide-button--${error ? 'expanded' : 'collapsed'}`} onClick={this.dismiss}/>
            </div>
          </div>
        )
      }

      return (
        <div className='Redux-Bug-Reporter'>
          {expanded && (
            <div className='Redux-Bug-Reporter__form'>
              <form onSubmit={this.submit}>
                <input className='Redux-Bug-Reporter__form-input Redux-Bug-Reporter__form-input--reporter' onChange={this.handleChange('reporter')} value={reporter} placeholder='Name'/>
                <input className='Redux-Bug-Reporter__form-input Redux-Bug-Reporter__form-input--description' onChange={this.handleChange('description')} value={description} placeholder='Description'/>
                <input className='Redux-Bug-Reporter__form-input Redux-Bug-Reporter__form-input--screenshotURL' onChange={this.handleChange('screenshotURL')} value={screenshotURL} placeholder='Screenshot URL'/>
                <textarea className='Redux-Bug-Reporter__form-input Redux-Bug-Reporter__form-input--notes' onChange={this.handleChange('notes')} value={notes} placeholder='Notes'/>
                <button className='Redux-Bug-Reporter__submit-button' type='submit'>File Bug</button>
              </form>
            </div>
          )}
          <div className='Redux-Bug-Reporter__show-hide-container'>
            <button className={`Redux-Bug-Reporter__show-hide-button Redux-Bug-Reporter__show-hide-button--${this.state.expanded ? 'expanded' : 'collapsed'}`} onClick={this.toggleExpanded}/>
          </div>
        </div>
      )
    }
  })
}

const loadingLayout = (
    <div className='Redux-Bug-Reporter'>
      <div className='Redux-Bug-Reporter__loading-container'>
        <span className='Redux-Bug-Reporter__loading'/>
      </div>
    </div>
)

const mapStateToProps = (store) => {
  return {
    storeState: store
  }
}

const mapDispatchToProps = (dispatch) => {
  const boundActions = bindActionCreators({overloadStore, initializePlayback, finishPlayback}, dispatch)
  return {
    dispatch,
    ...boundActions
  }
}

const ConnectedBugReporter = connect(mapStateToProps, mapDispatchToProps)(UnconnectedBugReporter)

export { UnconnectedBugReporter }
export default ConnectedBugReporter
