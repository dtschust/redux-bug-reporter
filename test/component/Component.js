/* global fetch */
import { createStore, applyMiddleware, compose } from 'redux'
import { fromJS } from 'immutable'
import thunk from 'redux-thunk'

import React from 'react'
import Enzyme, { mount, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { Provider } from 'react-redux'
import { errorData } from '../../src/utils'
import {storeEnhancer} from '../../src/index'

import ReduxBugReporter, { UnconnectedBugReporter } from '../../src/redux-bug-reporter'

Enzyme.configure({ adapter: new Adapter() })
global.fetch = require('jest-fetch-mock')

require('es6-promise').polyfill()
require('isomorphic-fetch')

function reducer (state = [], action) {
  if (action) {
    return [...state, action]
  }
  return state
}

function configureStore (initialState) {
  return createStore(reducer, initialState, compose(storeEnhancer, applyMiddleware(thunk)))
}

function immutableReducer (state = fromJS([]), action) {
  if (action) {
    return state.push(action)
  }
  return state
}

function configureImmutableStore (initialState) {
  return createStore(immutableReducer, initialState, compose(storeEnhancer, applyMiddleware(thunk)))
}

describe('Redux Bug Reporter component tests', () => {
  beforeEach(() => {
    fetch.resetMocks()
  })

  let reduxBugReporter
  const udpatedReduxBugReporter = wrapper => wrapper
    .update()
    .find(UnconnectedBugReporter)

  it('Happy path', (done) => {
    fetch.mockResponse(JSON.stringify({
      bugURL: 'http://redux-bug-reporter.com/id/1'
    }), 200)
    const store = configureStore()
    const wrapper = mount(
      <Provider store={store}>
        <ReduxBugReporter
          projectName='foo'
          submit='http://redux-bug-reporter.com'
        />
      </Provider>
    )
    reduxBugReporter = udpatedReduxBugReporter(wrapper)
    const showHideButton = wrapper.find('.Redux-Bug-Reporter__show-hide-button')
    expect(reduxBugReporter.find('.Redux-Bug-Reporter')).toBeTruthy()

    // Form not initially expanded
    expect(reduxBugReporter.find('.Redux-Bug-Reporter__form').length).toEqual(0)
    showHideButton.simulate('click')
    reduxBugReporter = udpatedReduxBugReporter(wrapper)

    // Expanded
    expect(reduxBugReporter.find('.Redux-Bug-Reporter__form').length).toEqual(1)
    showHideButton.simulate('click')
    reduxBugReporter = udpatedReduxBugReporter(wrapper)

    // Collapsed
    expect(reduxBugReporter.find('.Redux-Bug-Reporter__form').length).toEqual(0)

    showHideButton.simulate('click')
    reduxBugReporter = udpatedReduxBugReporter(wrapper)

    // Expanded
    expect(reduxBugReporter.find('.Redux-Bug-Reporter__form').length).toEqual(1)

    // Edit the inputs
    // TODO: Validate inputs
    reduxBugReporter.find('.Redux-Bug-Reporter__form-input--reporter').simulate('change', { target: { value: 'Drew Schuster' } })
    reduxBugReporter = udpatedReduxBugReporter(wrapper)

    // Submit the bug
    const form = wrapper.find('form')
    form.prop('onSubmit')({ preventDefault: () => {} })

    // Loading
    reduxBugReporter = udpatedReduxBugReporter(wrapper)
    expect(reduxBugReporter.find('.Redux-Bug-Reporter__loading-container').length).toEqual(1)

    // Bug submission complete
    setTimeout(() => {
      expect(fetch.mock.calls[0][0]).toEqual('http://redux-bug-reporter.com')
      expect(fetch.mock.calls[0][1].method).toEqual('post')
      reduxBugReporter = udpatedReduxBugReporter(wrapper)
      expect(reduxBugReporter.find('.Redux-Bug-Reporter__form--success').length).toEqual(1)
      expect(reduxBugReporter.find('.Redux-Bug-Reporter__form--success').html()).toContain('http://redux-bug-reporter.com/id/1')

      const closeButton = reduxBugReporter.find('button')
      closeButton.simulate('click')
      reduxBugReporter = udpatedReduxBugReporter(wrapper)
      expect(reduxBugReporter.find('.Redux-Bug-Reporter')).toBeTruthy()
      expect(reduxBugReporter.find('.Redux-Bug-Reporter__form').length).toEqual(0)
      done()
    }, 0)
  })
  it('customEncode and customDecode properties', (done) => {
    fetch.mockResponse(JSON.stringify({
      bugURL: 'http://redux-bug-reporter.com/id/1'
    }), 200)
    const customEncode = (state) => state.toJSON()
    const customDecode = (state) => fromJS(state)
    const store = configureImmutableStore()
    const wrapper = mount(
      <Provider store={store}>
        <ReduxBugReporter
          projectName='foo'
          submit='http://redux-bug-reporter.com'
          customEncode={customEncode}
          customDecode={customDecode}
        />
      </Provider>
    )
    store.dispatch({type: 'FOO'})
    reduxBugReporter = udpatedReduxBugReporter(wrapper)
    const showHideButton = wrapper.find('.Redux-Bug-Reporter__show-hide-button')

    // Is rendered
    expect(reduxBugReporter.find('.Redux-Bug-Reporter')).toBeTruthy()

    // Form not initially expanded
    expect(reduxBugReporter.find('.Redux-Bug-Reporter__form').length).toEqual(0)
    showHideButton.simulate('click')
    reduxBugReporter = udpatedReduxBugReporter(wrapper)

    // Expanded
    expect(reduxBugReporter.find('.Redux-Bug-Reporter__form').length).toEqual(1)
    showHideButton.simulate('click')
    reduxBugReporter = udpatedReduxBugReporter(wrapper)

    // Collapsed
    expect(reduxBugReporter.find('.Redux-Bug-Reporter__form').length).toEqual(0)

    showHideButton.simulate('click')
    reduxBugReporter = udpatedReduxBugReporter(wrapper)

    // Expanded
    expect(reduxBugReporter.find('.Redux-Bug-Reporter__form').length).toEqual(1)

    // Edit the inputs
    // TODO: Validate inputs
    reduxBugReporter.find('.Redux-Bug-Reporter__form-input--reporter').simulate('change', { target: { value: 'Drew Schuster' } })

    // Submit the bug
    const form = wrapper.find('form')
    form.prop('onSubmit')({ preventDefault: () => {} })
    reduxBugReporter = udpatedReduxBugReporter(wrapper)

    // loading
    expect(reduxBugReporter.find('.Redux-Bug-Reporter__loading-container').length).toEqual(1)

    // Bug submission complete
    setTimeout(() => {
      expect(fetch.mock.calls[0][0]).toEqual('http://redux-bug-reporter.com')
      expect(fetch.mock.calls[0][1].method).toEqual('post')
      reduxBugReporter = udpatedReduxBugReporter(wrapper)
      expect(reduxBugReporter.find('.Redux-Bug-Reporter__form--success').length).toEqual(1)
      expect(reduxBugReporter.find('.Redux-Bug-Reporter__form--success').html()).toContain('http://redux-bug-reporter.com/id/1')

      const closeButton = reduxBugReporter.find('button')
      closeButton.simulate('click')
      reduxBugReporter = udpatedReduxBugReporter(wrapper)
      expect(reduxBugReporter.find('.Redux-Bug-Reporter')).toBeTruthy()
      expect(reduxBugReporter.find('.Redux-Bug-Reporter__form')).toBeTruthy()
      done()
    }, 0)
  })
  it('Custom submit function', (done) => {
    fetch.mockResponse(JSON.stringify({
      bugURL: 'http://redux-bug-reporter.com/id/2'
    }), 200)
    const store = configureStore()
    const submitFn = (newBug) => fetch('http://redux-bug-reporter.com', {
        method: 'post',
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(newBug)
      }).then((response) => {
        if (!response.ok) {
          throw Error(response.statusText)
        }
        return response.json()
      })
    const wrapper = mount(
      <Provider store={store}>
        <ReduxBugReporter
          projectName='foo'
          submit={submitFn}
        />
      </Provider>
    )
    reduxBugReporter = udpatedReduxBugReporter(wrapper)
    const showHideButton = wrapper.find('.Redux-Bug-Reporter__show-hide-button')

    // Rendered
    expect(reduxBugReporter.find('.Redux-Bug-Reporter')).toBeTruthy()

    showHideButton.simulate('click')
    reduxBugReporter = udpatedReduxBugReporter(wrapper)

    // Expands
    expect(reduxBugReporter.find('.Redux-Bug-Reporter__form').length).toEqual(1)

    // Submit the bug
    const form = wrapper.find('form')
    form.prop('onSubmit')({ preventDefault: () => {} })
    reduxBugReporter = udpatedReduxBugReporter(wrapper)

    expect(reduxBugReporter.find('.Redux-Bug-Reporter__loading-container').length).toEqual(1)

    // Bug submission complete
    setTimeout(() => {
      expect(fetch.mock.calls[0][0]).toEqual('http://redux-bug-reporter.com')
      expect(fetch.mock.calls[0][1].method).toEqual('post')
      reduxBugReporter = udpatedReduxBugReporter(wrapper)
      expect(reduxBugReporter.find('.Redux-Bug-Reporter__form--success').length).toEqual(1)
      expect(reduxBugReporter.find('.Redux-Bug-Reporter__form--success').html()).toContain('http://redux-bug-reporter.com/id/2')
      done()
    }, 0)
  })
  it('Error listeners', () => {
    errorData.clearErrors()
    const store = configureStore()
    mount(
      <Provider store={store}>
        <ReduxBugReporter
          projectName='foo'
          submit='http://redux-bug-reporter.com'
        />
      </Provider>
    )

    // verify overrides
    expect(window.console.error.bugReporterOverrideComplete).toBeTruthy()
    expect(window.onerror.bugReporterOverrideComplete).toBeTruthy()

    // normal console error
    window.console.error('Here is a console error')

    expect(errorData.getErrors()).toEqual([{ errorMsg: 'Here is a console error' }])

    // console error with stack trace
    window.console.error({
      name: 'name',
      message: 'here is a message',
      stack: 'This is a stack trace'
    })
    let expected = {
      errorMsg: 'name: here is a message',
      stackTrace: 'This is a stack trace'
    }

    expect(errorData.getErrors()[1]).toEqual(expected)

    // window onerror
    expected = {
      errorMsg: 'simple error message',
      stackTrace: undefined
    }
    window.onerror('simple error message')
    expect(errorData.getErrors()[2]).toEqual(expected)

    // window onerror with stack trace
    expected = {
      errorMsg: 'message',
      stackTrace: 'stack trace'
    }
    window.onerror('message', undefined, undefined, undefined, { stack: 'stack trace' })
    expect(errorData.getErrors()[3]).toEqual(expected)
  })
  it('window.onerror listeners when window.onerror already exists', () => {
    errorData.clearErrors()
    // make window.onerror exist
    let originalCalled = false
    window.onerror = function onerror() {
      originalCalled = true
    }
    const store = configureStore()
    mount(
      <Provider store={store}>
        <ReduxBugReporter
          projectName='foo'
          submit='http://redux-bug-reporter.com'
        />
      </Provider>
    )

    // verify overrides
    expect(window.onerror.bugReporterOverrideComplete).toBeTruthy()

    // window onerror
    let expected = {
      errorMsg: 'simple error message',
      stackTrace: undefined
    }
    window.onerror('simple error message')
    expect(errorData.getErrors()[0]).toEqual(expected)
    expect(originalCalled).toBeTruthy()

    // window onerror with stack trace
    originalCalled = false
    expected = {
      errorMsg: 'message',
      stackTrace: 'stack trace'
    }
    window.onerror('message', undefined, undefined, undefined, { stack: 'stack trace' })
    expect(errorData.getErrors()[1]).toEqual(expected)
    expect(originalCalled).toBeTruthy()
  })
})

describe('Redux Bug Reporter playback tests', () => {
  let defaultArgs
  let defaultProps
  const spyConsoleLog = jest
    .spyOn(console, 'log')
    .mockImplementation(() => {})

  beforeEach(() => {
    defaultArgs = {
      actions: [{}],
      initialState: null,
      finalState: null,
      delay: undefined,
    }
    defaultProps = {
      customDecode: jest.fn(arg => arg),
      dispatch: jest.fn(),
      finishPlayback: jest.fn(),
      initializePlayback: jest.fn(),
      overloadStore: jest.fn(),
      projectName: 'foo',
      storeState: {},
      submit: 'http://redux-bug-reporter.com',
    }
  })

  afterEach(() => {
    defaultProps.customDecode.mockReset()
    defaultProps.dispatch.mockReset()
    defaultProps.finishPlayback.mockReset()
    defaultProps.initializePlayback.mockReset()
    defaultProps.overloadStore.mockReset()
    spyConsoleLog.mockReset()
  })

  it('aborts when delay === -1', () => {
    const props = { ...defaultProps }
    const args = { ...defaultArgs, delay: -1 }
    const instance = shallow(<UnconnectedBugReporter { ...props } />).instance()
    instance.bugReporterPlayback(...Object.values(args))
    expect(instance.props.overloadStore).toHaveBeenCalledTimes(1)
  })

  it('makes function calls', () => {
    const props = { ...defaultProps }
    const args = { ...defaultArgs }
    const instance = shallow(<UnconnectedBugReporter { ...props } />).instance()
    instance.bugReporterPlayback(...Object.values(args))
    const { customDecode, initializePlayback, overloadStore } = instance.props
    expect(customDecode).toHaveBeenCalledTimes(2)
    expect(initializePlayback).toHaveBeenCalledTimes(1)
    expect(overloadStore).toHaveBeenCalledTimes(1)
    expect(spyConsoleLog).toHaveBeenCalledTimes(1)
  })

  it('does not call `customDecode` when undefined', () => {
    const props = { ...defaultProps, customDecode: undefined }
    const args = { ...defaultArgs }
    const instance = shallow(<UnconnectedBugReporter { ...props } />).instance()
    instance.bugReporterPlayback(...Object.values(args))
    const { customDecode } = instance.props
    expect(customDecode).toBe(undefined)
  })

  it('iterates actions', () => {
    const props = { ...defaultProps }
    const args = {
      ...defaultArgs,
      actions: [
        defaultArgs.actions[0],
        { type: 'ACTION_ONE', payload: { property: 'ACTION_ONE_PAYLOAD_PROPERTY' } },
        { type: 'ACTION_TWO', payload: { property: 'ACTION_TWO_PAYLOAD_PROPERTY' } },
      ],
    }
    const instance = shallow(<UnconnectedBugReporter { ...props } />).instance()
    instance.bugReporterPlayback(...Object.values(args))
    const { customDecode, dispatch, initializePlayback, overloadStore } = instance.props
    expect(customDecode).toHaveBeenCalledTimes(2)
    expect(dispatch).toHaveBeenCalledTimes(1)
    expect(initializePlayback).toHaveBeenCalledTimes(1)
    expect(overloadStore).toHaveBeenCalledTimes(1)
    expect(spyConsoleLog).toHaveBeenCalledTimes(0)
  })

  it('reports divergence between storeState and finalState', () => {
    const storeState = { slice: { property: 'SLICE_PROPERTY' } }
    const finalState = { slice: { mismatch: 'SLICE_MISMATCH' } }
    const props = { ...defaultProps, storeState }
    const args = { ...defaultArgs, finalState }
    const instance = shallow(<UnconnectedBugReporter { ...props } />).instance()
    instance.bugReporterPlayback(...Object.values(args))
    expect(spyConsoleLog).toHaveBeenCalledTimes(3)
  })

  it('does not report when storeState and finalState are the same', () => {
    const storeState = { slice: { property: 'SLICE_PROPERTY' } }
    const finalState = { slice: { property: 'SLICE_PROPERTY' } }
    const props = { ...defaultProps, storeState }
    const args = { ...defaultArgs, finalState }
    const instance = shallow(<UnconnectedBugReporter { ...props } />).instance()
    instance.bugReporterPlayback(...Object.values(args))
    expect(spyConsoleLog).toHaveBeenCalledTimes(1)
  })
})

describe('Redux Bug Reporter render', () => {
  let defaultProps

  beforeEach(() => {
    defaultProps = {
      customDecode: () => {},
      dispatch: () => {},
      finishPlayback: () => {},
      initializePlayback: () => {},
      overloadStore: () => {},
      projectName: 'foo',
      storeState: {},
      submit: 'http://redux-bug-reporter.com',
    }
  })

  it('renders button without the form', () => {
    const props = { ...defaultProps }
    const wrapper = shallow(<UnconnectedBugReporter { ...props } />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders button and form', () => {
    const props = { ...defaultProps }
    const wrapper = shallow(<UnconnectedBugReporter { ...props } />)
    wrapper.instance().setState({ expanded: true })
    wrapper.update()
    expect(wrapper).toMatchSnapshot()
  })

  it('renders error message', () => {
    const props = { ...defaultProps }
    const wrapper = shallow(<UnconnectedBugReporter { ...props } />)
    wrapper.instance().setState({
      bugFiled: true,
      error: true,
    })
    wrapper.update()
    expect(wrapper).toMatchSnapshot()
  })

  it('renders success after filing bug report', () => {
    const props = { ...defaultProps }
    const wrapper = shallow(<UnconnectedBugReporter { ...props } />)
    wrapper.instance().setState({
      bugFiled: true,
      bugURL: 'BUG_URL',
    })
    wrapper.update()
    expect(wrapper).toMatchSnapshot()
  })
})

// TODO: Test submission failure
// TODO: Test actually filling out the form, something's hard to mock out there
// TODO: Test redactstorestate prop
// TODO: Somehow test not mounted
// TODO: Props to test: name, projectName
// TODO: Verify that real network requests aren't made during playback
// TODO: Test meta
