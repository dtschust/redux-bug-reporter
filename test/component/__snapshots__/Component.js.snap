// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redux Bug Reporter render renders button and form 1`] = `
<div
  className="Redux-Bug-Reporter"
>
  <div
    className="Redux-Bug-Reporter__form"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        className="Redux-Bug-Reporter__form-input Redux-Bug-Reporter__form-input--reporter"
        onChange={[Function]}
        placeholder="Name"
        value=""
      />
      <input
        className="Redux-Bug-Reporter__form-input Redux-Bug-Reporter__form-input--description"
        onChange={[Function]}
        placeholder="Description"
        value=""
      />
      <input
        className="Redux-Bug-Reporter__form-input Redux-Bug-Reporter__form-input--screenshotURL"
        onChange={[Function]}
        placeholder="Screenshot URL"
        value=""
      />
      <textarea
        className="Redux-Bug-Reporter__form-input Redux-Bug-Reporter__form-input--notes"
        onChange={[Function]}
        placeholder="Notes"
        value=""
      />
      <button
        className="Redux-Bug-Reporter__submit-button"
        type="submit"
      >
        File Bug
      </button>
    </form>
  </div>
  <div
    className="Redux-Bug-Reporter__show-hide-container"
  >
    <button
      className="Redux-Bug-Reporter__show-hide-button Redux-Bug-Reporter__show-hide-button--expanded"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`Redux Bug Reporter render renders button without the form 1`] = `
<div
  className="Redux-Bug-Reporter"
>
  <div
    className="Redux-Bug-Reporter__show-hide-container"
  >
    <button
      className="Redux-Bug-Reporter__show-hide-button Redux-Bug-Reporter__show-hide-button--collapsed"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`Redux Bug Reporter render renders error message 1`] = `
<div
  className="Redux-Bug-Reporter"
>
  <div
    className="Redux-Bug-Reporter__form Redux-Bug-Reporter__form--fail"
  >
    <div>
      <div>
        Oops, something went wrong!
      </div>
      <div>
        Please try again later
      </div>
    </div>
  </div>
  <div
    className="Redux-Bug-Reporter__show-hide-container"
  >
    <button
      className="Redux-Bug-Reporter__show-hide-button Redux-Bug-Reporter__show-hide-button--expanded"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`Redux Bug Reporter render renders success after filing bug report 1`] = `
<div
  className="Redux-Bug-Reporter"
>
  <div
    className="Redux-Bug-Reporter__form Redux-Bug-Reporter__form--success"
  >
    <div>
      <div>
        Your bug has been filed successfully!
      </div>
      <div>
        <a
          href="BUG_URL"
          target="_blank"
        >
          Here is a link to it!
        </a>
      </div>
    </div>
  </div>
  <div
    className="Redux-Bug-Reporter__show-hide-container"
  >
    <button
      className="Redux-Bug-Reporter__show-hide-button Redux-Bug-Reporter__show-hide-button--collapsed"
      onClick={[Function]}
    />
  </div>
</div>
`;
